### PR TITLE
Sound types for the `addRoute'-family of functions.

### PR DESCRIPTION
**Pull request by dhil**
This patch refines the type of `unsafeAddRoute`, and its relatives, in
particular `addRoute`, which previously had the type

    addRoute : (String, (String) ~> Page) ~> ()

which elaborates to

    addRoute : (String, (String) ~e~> Page) ~f~> ()

where `e` and `f` are distinct effect variables. Consequently the
invariant that all user-definable effectful operations are handled
prior to reaching the toplevel is broken, for instance

    addRoute("/", fun(_) {do Fail})

would cause a "no handler installed for Fail" error to occur at
runtime when visiting the root page, even though one might have
wrapped the `addRoute` in a handler, i.e.

    handle(addRoute("/", fun(_) {do Fail})) {
      case Fail -> ()
    }

This does not handle the effects of the anonymous function, because the function
gets executed in a separate context where the handler is not
installed.

To ensure that all user-definable computational effects are handled,
this patch "corrects" the type of `unsafeAddRoute` and its associated
functions, as a result the type of `addRoute` is now

    addRoute : (String, (String) {hear{_}}~> Page) ~> ()

which elaborates to

    addRoute : (String, (String) {hear{p}}~> Page) ~f~> ()

In other words, a page generating function must handle all of its
effects -- save for the two built-in effects `hear` and `wild`.

### CODE REVIEW

**Comment by dhil** [comment]
> I have a slightly bigger concern however. I tried your original example (but with servePages added afterwards, so I could try it):

I don't see how that example can type check with this patch, and indeed on my branch it fails to type check.

**Comment by SimonJF** [comment]
Apologies; I was on the wrong branch (d'oh, more coffee needed). I've rechecked everything and it's fine. Thanks!

Please add those unit tests, as well as 

```
handle({ addRoute("/", fun(_) {do Fail}); servePages() }) {
  case Fail -> debug("failed")
}
```

and I think we're good to go.

**Comment by SimonJF** [comment]
Well spotted, and thanks for the patch!
